### ISSUE TITLE
Better build system

### ISSUE THREAD

**Issue by b-g** [hfg-gmuend/openmoji#154]
Hi @kriskowal and @carlinmack,

Wow!! Many thanks for the PR!! As they are all build system related (and somehow conflicting each other) I thought we should bundle the discussion here. I hope you don't mind.

Comments:

@kriskowal 
- #149 this is a draft and if currently not intended for being merged, or? I guess we won't be able to merge it until your bug fix for SVGO was accepted. 
- #151 and #152 is ready to be merged, correct? 

@carlinmack 
- #153 Many thanks ... but this seems to be already solved with the PRs above.

Hmm ... I have to say I extremely appreciate the PRs, but I'm a bit overwhelmed by them. I think we (Daniel, Hartmut and me) are less experts in the field of dev ops and it is quite difficult for us to evaluate the implications of accepting them. After all I will have to maintain OpenMoji with code which I currently don't fully understand. This will create kind of a new dependency to @kriskowal :)

Hence may I suggest the following:

1.
- we create a branch `dev-build-system`; add you both as contributors
- we merge #151 #152 to branch `dev-build-system`
- we all test it
- iterate on it together
- once we are all fine with -> merge it with master

2.
Afterwards ... same process for #149 (if SVGO is patched) **or** we add a temporary feature e.g. `export single hexcode` as propsed in #153 

Is this an approach the two of you could support/accept?

**Comment by carlinmack** [comment]
I'm not sure that @kriskowal s PR will solve my problem as I seem to generate different files that are visually identical with the current export scripts. I think the caching of input technique will maybe help this though.

**Comment by kriskowal** [comment]
#149 is an earnest proposal to speed up incremental pretty-src-svg and does indeed require an svgo bugfix before it’s ready for review.

#151 is also an earnest proposal, but loaded with bash style guide opinions worth review. Since it touches all the scripts, it’s likely to cause conflicts. It may be better as a reference if you continue to evolve other scripts. I do think it will be helpful to add shellcheck to CI before larger sweeping changes since it establishes a good baseline for style.

#152 is merge-able but I believe @carlinmack and I are both continuing to experiment on further improvement. I agree being able to pass codes as variadic arguments to focus the build is useful. I also have an experiment open to add caching. To get good performance from caching, I’m taking a more novel approach than I did with pretty-svg-src memo-in-place. It’s possible that this approach will be fast enough for incremental builds that it provides a better developer experience than specifying codes as arguments.

Regardless, these are gifts you’re free to drop. I cut #152 early since it seems to strike a good compromise between speed and maintainability (read: not necessarily adding a dependency on me for all further review). The optimizations I’m puttering with today are more esoteric.

**Comment by kriskowal** [comment]
I’ve added #156 to the pile of options to consider for export-png. This one is more elaborate and I would advise only using it if another core maintainer can make sense of it. I believe it provides the best user experience of the options. Both bulk and incremental builds are quite fast.

**Comment by b-g** [comment]
@kriskowal Hi! Started to look into it! Tested and merged #151 the "script review" PR ... the helper scripts look much nicer now! Many thanks! Will be a good template for future bash helpers. Will continue testing and merging over the course of the next days.

Just a quick question ... can you elaborate on these lines? What are doing?

https://github.com/hfg-gmuend/openmoji/blob/52c202feb4d0fcd826bd2936a9afb1b4635baa97/helpers/export-emojipedia.sh#L1-L3

**Comment by kriskowal** [comment]
This is a cocktail of shell script best practices.

`set -u` causes any reference to an undefined environment variable to abort the process with an error. The default behavior is to interpret an undefined variable as an empty string. This setting makes it easier to detect variable name typos and similar defects.

`set -e` causes bash to exit immediately with the status code of the first statement with a nonzero exit status code. This allows exceptions to propagate instead of ignoring them.

`set -o pipefail` causes the process pipelines to propagate errors from any of the underlying processes, not just the last one.

This appears to be a good explainer for the options: https://sipb.mit.edu/doc/safe-shell/

`IFS=$'\t\n'` sets the input field separator that many bash expressions, builtins, and commands use to split and join input fields. Bash uses the first character to join fields and uses any of the characters to split. The default includes space, which tends to cause surprises, particularly when file names contain spaces.

For example, `export IFS=','; echo "$*"` will echo arguments joined by `,`. The `read` command by default reads input until the next encountered IFS.

I don’t believe any of these scripts are particularly sensitive to IFS, but it’s good to have sane defaults if the script evolves. Some of the defects a bad IFS can cause can easily go undetected until the input evolves to include a space somewhere.

**Comment by b-g** [comment]
Dear @kriskowal,

I've spent some time trying to understand your better build system proposals. Basically I think everything in terms of bit and pieces is already great, but I would architecture if slightly different. Below is an attempts to document it somehow.

`helpers/generate.sh`
```
1. find all changed svg files in the src folder via a git memo table.

2. process only "changed src files" and generate all dependent assets in place. 
Pseudo code:
for src_svg in changed_src_svg
    thread( node src-to-svg-black(src_svg, black/svg) )
    thread( node src-to-svg-color(src_svg, color/svg) )
    ... 
    thread( svg-to-png(svg_color, color/72x72) )
    thread( svg-to-png(svg_color, color/618x618) )
    ...
    thread( memo( beautify-svg(src_svg) ) )
```

And refactor the reusable commands/functions into a `helper/lib` folder
```
helpers/lib/src-to-svg-black.js
helpers/lib/src-to-svg-color.js
helpers/lib/beautify-svg.sh
helpers/lib/thread.sh
helpers/lib/memo.sh
helpers/lib/svg-to-png.sh
helpers/lib/svg-to-pdf.sh
helpers/lib/create-zip.sh
```

This would replace then over time most export-something.sh script and the node.js based script to fit the new memo approach.

What do you think?

**Comment by kriskowal** [comment]
From #163, I demonstrated for myself that the steps I proposed for moving away from checked-in generated assets have some wrinkles.

Unless openmoji introduces a development dependency on Docker (which would certainly be hostile to designers), it really isn’t practical to hope that the assets generated for each individual PR will be identical or representative of the assets that would be released. This tells me that `generate.sh` is really only for release management, not development. This also means that in the short term, developers should be using a workflow like @carlinmack is proposing with #153 to generate assets for a specific source file is the right direction for the project.

As written, the changes I’ve proposed for doing bulk cached incremental builds wouldn’t be useful for development since they would cause a lot of thrashing around assets that are unrelated to proposed changes. They would certainly be helpful for generating releases. If they’re fast enough, they could be used for CI/CD and eliminate the need for checking in generated assets. Regardless, it would be handy if release management produced consistent artifacts from different workstations.

Your suggestion above about separating helpers/lib out to make it clear which scripts are user-facing is good. My proposed changes would need to be converted to have a mode where they only rebuild specific hex codes like @carlinmack’s change to be useful in development. I think the ideal user-experience would be for helpers/generate.sh to be the only user-facing script, and for it to accept the hex code or source file under development (or multiple), and generate all of the targets from that source or sources, or generate everything if no arguments are specified.

Is this a sensible direction?

**Comment by b-g** [comment]
Hi @kriskowal and @carlinmack, many thanks for all thoughts, suggestions, tests and proposals. I really appreciate!! But I find it hard to follow the discussion across PRs, Github Issues and emails :) Hence could we discuss in a quick zoom / skype / hangout in person? As I currently have my two small boys (no kindergarten in times of COVID) I'm a bit limited in terms of time slots. What about?

- April Wed 8: 5p or 6p CEST
- April Thu 9: 5p or 6p CEST
- If so ... may I send you the confirmation and invites via twitter DM?
- Any strong preferences in terms of zoom / skype / hangout?

**Comment by kriskowal** [comment]
Sorry for the late reply. I could be available for a 6pm CEST / 8am PDT call (in a half hour from now).

My email is kris@cixar.com. My DM’s are closed so I don’t believe that will work.

Zoom works best for me. I’ll drop a meeting link here if I don’t hear from you.

**Comment by kriskowal** [comment]
I’ll be here if you happen to still be available.

https://zoom.us/j/528889104?pwd=N0kyNjQ4S3B1ZFZuR2haNnlrREhFUT09

**Comment by b-g** [comment]
Dear @kriskowal! Many thanks for the generous gift of #156!! Please give me a couple of days to check / test / understand it. Will get back to you on this latest mid of next week.

Off topic: We ask everyone who contributed an OpenMoji to submit a self portrait (and potential social media links), so that we can feature you at https://openmoji.org/about/#team = the hall of fame :) This is not a must ... rather a lovely tradition most contributors followed. If so, you can either post the portrait/infos here or directly do a PR at https://github.com/hfg-gmuend/openmoji-website. Many thanks for considering! 🙏

**Comment by kriskowal** [comment]
Of course. Let me know if it glitches in any particular state.

I would be flattered if you would use: https://openmoji.org/library/#search=mage&emoji=1F9D9-1F3FB-200D-2642-FE0F I believe you know where to find me on Twitter and Github.

**Comment by b-g** [comment]
@kriskowal Added your credits https://github.com/hfg-gmuend/openmoji-website/commit/67e68a6693cf786ca564ef320d38ef7808060d7f :)

**Comment by b-g** [comment]
Hi @kriskowal, 

Finally had time to unwrap your gift #156! Very nice!! 

Comments:

- We are on node 8.x, hence I had to add a shim. Please check my edit. The line looks quite unreadable now. Could you unpack this particular line? https://github.com/hfg-gmuend/openmoji/blob/buildsystem-with-cache/helpers/lib/export-svg-color.js#L27
- Would it not be better to move helpers/export-svg-skintones.sh, helpers/export-svg-color.sh, helpers/export-svg-black.sh to the lib folder? I don't think one should call them independent from each other.

Also I've edited a few tiny other things, please continue based on my branch https://github.com/hfg-gmuend/openmoji/tree/buildsystem-with-cache

Would test it a bit further in the next days and wait for potential edits of yours ... and then merge it. OK?

**Comment by kriskowal** [comment]
I’ve posted a tiny amendment. Instead of adding an external dependency on an Object.fromEntries shim all its transitive dependencies, I elected to reframe the lines with fromEntries as a more readable loop, instead of the functional form. The functional composition stuff you can do in newer JS versions is nice, but not worth taking on a dependency for. I would go so far as to suggest removing the lodash dependency: much of its utility is built-into the language now.

I also consolidated all the export-svg scripts into the single entrypoint export-svg.sh script. They are order dependent and generally fast enough that I don’t think folks will need to walk through them manually anymore. If need be, we can separate them into lib/ versions and rename some files according to whether they’re plan-phase or execute-phase.

**Comment by b-g** [comment]
@kriskowal Finally merged! :) Added some documentation and tweaked the font generator script. Yay! Really super great to have this in place now! Many many thanks for this gift! 🙏!!
